Export code static labeling to documentation

katie-lamb opened the pull request:
A rough draft on exporting the codes dataframes to documentation. Notable questions/changes/issues include:  

- Restructured the dictionaries in `metadata.codes.py` into one dictionary called `CODES_METADATA` mainly because it made iterating through and using each of the codes easier. Not sure if this was really the thing to do. It got a little messy.
- The codes dataframes are made into CSV's before being rendered in the RST format. I checked the CSV directory in for the sake of debugging but ultimately think this should probably be a temp or cleaned up directory
- I need to clean up the jinja template to handle paragraph blocks better in the descriptions
- `tox -e docs` is failing for me (in dev and main) because of an issue importing `addfips` which seems like it might be a new issue that maybe I remember Trenton working on. Or maybe it's something to do with the new pudl-dev environment.
- I was generally confused about Sphinx and wanted to see if the tox tests still passed so I commented out the call to`static_dfs_to_rst` in the setup in `conf.py`. Not entirely sure this was even the right place to call it. Need to look more into this

General To Do

- [x] Figure out how Sphinx works and call `static_dfs_to_rst` during build
- [x] Clean up CSV directory after the RST has been created
- [x] Clean up strings in CODE_DESCRIPTIONS dictionary (need to better handle paragraph breaks in RST file)
- [x] Need an assertion for the new Encoder constructor in case a code name is passed that is not in the `CODES_METADATA` dictionary?
- [x] Move `description` and `doc_name` field in Encoder - use Resource instead and the encoder field inside of Resource? Resource isn't necessarily built to just hold one field/code? - UPDATE: Resource is pretty equipped to hold `description`, `doc_name`, and `encoder` already but it already has a `to_rst` method so I chose to use a list of `Encoders` instead.
- [x] Remove the generated `codes_and_labels.rst` file in `cleanup_rst` once the build is finished? 

Anyways, just wanted to get some eyes on it because I felt like a lot of stuff happened in the process of me learning about the docs.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -1632,3 +1633,45 @@ def to_sql(
+                description = None
+                title_name = name
+            csv_filepath = os.path.join(csv_dir, f"{name}.csv")
+            CODE_METADATA[name]["df"].to_csv(csv_filepath)
```
Outputting the CSV as a side effect of instantiating the classes seems a little sketchy / magical. It seems like all the data should be stored within the class itself, and then the `to_rst()` method should deal with the mechanics of exporting the CSV to wherever it needs to go -- maybe have `to_rst()` automatically create a directory full of CSV files next to the RST file which is being generated, that is referenced from within the generated RST?

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -1632,3 +1633,45 @@ def to_sql(
         return metadata
+
+
+class CodeData(Base):
```
Collecting all of the coding tables into a single data structure that we can refer to and iterate over makes sense to me. It's stored as a dictionary. How does it benefit from being a list here instead?

In the `Resource` and `Field` classes, we're using `from_XXX_ids()` rather than `from_XXX_names()` but they're doing the same thing, so I think we should use the same naming convention for the class methods.

There's already a Pydantic class defined here (`Encoder`) to contain the coding table information. Would it make sense to have this class contain a collection of `Encoder` objects instead? They'd get validated in the process too.

Would it be cleaner to add a `to_rst()` method to the `Encoder` class so that it can generate its own RST snippet? It would also need to know the directory to which the CSVs should be written so it can output the CSV file to generate the table from in the right place, and correctly reference that path in the RST it outputs.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -1632,3 +1633,45 @@ def to_sql(
+            code_list.append(code_dict)
+        return cls(code_list=code_list)
+
+    def to_rst(self, path: str) -> None:
```
If this class contains a list of `Encoders` each of which knows has a `to_rst()` method then I think the `to_rst()` method here could just iterate over all of them calling their own `to_rst()` methods, and add whatever preamble / postscript / glue is required to turn them into a complete page.

zaneselvans replied:
Could also have a `cleanup()` method that's called to remove the CSV files after the docs build has completed.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -1632,3 +1633,45 @@ def to_sql(
+
+    @classmethod
+    def from_code_names(
+        cls, code_names: Iterable[str], csv_dir: String
```
Since we're passing in a `pathlib.Path` for `csv_dir` does it make sense to just keep it as a Path rather than casting to a String and needing to turn it back into some kind of Path?  It looks like Pydantic [has built-in support](https://pydantic-docs.helpmanual.io/usage/types/#pydantic-types) for Path, FilePath, and DirectoryPath types.

zaneselvans reviewed (changes requested):
Hey I think the general direction here looks good.

Let's talk about maybe using the existing Encoder class to contain some of this functionality, and the Resource class for some of the table-level metadata.

zaneselvans commented on src/pudl/metadata/templates/codedata.rst.jinja:
```diff
@@ -0,0 +1,19 @@
+{% for code in CodeData.code_list %}
+
+  | **{{ code.title_name }}**
+  | PUDL Field Name\: {{ code.var_name }}
```
I think it might be good to also display at least the fixed codes in the docs -- so people can see how we are re-mapping the originally reported data to the canonical values.

katie-lamb commented on src/pudl/metadata/codes.py:
```diff
@@ -4,234 +4,248 @@
 
-The purchase type under which fuel receipts occurred in the reporting month.
-"""
+CODE_DESCRIPTIONS = {
```
It seems like descriptions in `RESOURCE_METADATA` are pretty scarce, for example `eia_energy_sources` has no description in `RESOURCE_METADATA` but had a pretty lengthy one in the codes metadata. One option is to put in some logic so we check for a description in `RESOURCE_METADATA` first but maybe this should just be standardized in some way.

I've changed the structure here so descriptions are part of the `CODES_METADATA` dictionary and are a new field in the Encoder class.

zaneselvans replied:
They're scarce but they shouldn't be -- every single table needs to have a description in the metadata for inclusion in our documentation and output to the various other platforms where we're publishing data (like Datasette, the Intake catalogs, etc). `RESOURCE_METADATA` is the place the descriptions should go, since every Encoder also corresponds to a database table. We just haven't gotten around to filling them all in. Actually I thought I just recently merged in a PR closing #1333 but... no. Sorry, these changes are on my `eia861-metadata` branch.

zaneselvans replied:
Let me run the Tox tests and see if everything is working and see if we can get it merged into `dev` to avoid this collision.

katie-lamb commented:
Main changes in latest commits: 
- Switched to using the Encoder class so now the CodeMetadata class basically holds a list of Encoders.
- Added `to_rst` method to the Encoder class. `to_rst` in the CodeMetadata class just iterates through the Encoders and calls `to_rst` for each of them.
- Added two variables to the Encoder class - `description` and `doc_name`. Instead of doing this, I thought about just using the Resource class which already has variables for `description`, `title`, and `encoder` but the Resource class already has its own `to_rst` method and generally is used to serve a different purpose. 
- I think I did the calls to `static_dfs_to_rst` and `cleanup` correctly with Sphinx in `conf.py` but it's hard to test because I don't have a fully functional environment on this borrowed computer.
- Added `code_fixes` dictionary into the RST.

Left To Do: 

- [ ] Fix the jinja template so that it generates a nicely formatted RST. There are some errors right now with the RST
- [ ] Get descriptions from `RESOURCE_METADATA` instead of writing them into `CODES_METADATA`
- [ ] Clean up block strings of the code descriptions (need to better handle paragraph breaks in RST file)

katie-lamb commented on src/pudl/metadata/resources/eia.py:
```diff
@@ -59,9 +56,18 @@
+        },
+        "encoder": CODE_METADATA["entity_types_eia"],
+        "sources": ["eia861"]
+    },
```
I added this into `RESOURCE_METADATA` since it wasn't included in the latest version of `dev` and `entity_types_eia` is now being included in the code metadata doc. Probably needs `foreign_key_rules`?

zaneselvans replied:
Because the `entity_type` fields are still such a heterogeneous mess (#1392), I left them a free form string for now. It's going to take significant work to get them all on the same page, and that was beyond the scope of the issue I was working on. So we aren't using it as a coding table / foreign key yet.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -1632,3 +1655,39 @@ def to_sql(
+        Construct a dictionary of code dictionaries containing the code name as it appears in the docs, a description, and file path to a CSV of the code label dataframe.
+
+        Args:
+            code_names: A list of Code PUDL identifiers, keys to entries in the CODE_METADATA dictionary.
```
I think you mean `code_ids` here? Keeping with the same naming convention across all of these methods will help them be more readable.

zaneselvans commented on src/pudl/metadata/templates/codemetadata.rst.jinja:
```diff
@@ -0,0 +1,29 @@
+*Also included is a table of non-standard codes that are mapped to canonical, standardized codes.*
+{% endif %}
+
+| **{{ Encoder.name }}**
```
For easier navigation / linking & consistency can we make each coding table a header like in the [PUDL DB Data DIctionary](https://catalystcoop-pudl.readthedocs.io/en/latest/data_dictionaries/pudl_db.html). I think the larger headers will help distinguish between the different coding tables too, since they have multiple sections (description, codes, fixes) which run together a bit now.

zaneselvans commented on docs/conf.py:
```diff
@@ -147,13 +149,33 @@ def metadata_to_rst(app):
+    abs_csv_dir_path.mkdir(parents=True, exist_ok=True)
+    rel_csv_dir_path = "/data_dictionaries/code_csvs"
+    codemetadata = CodeMetadata.from_code_ids(sorted(CODE_METADATA.keys()))
+    codemetadata.to_rst(abs_csv_dir_path=abs_csv_dir_path, rel_csv_dir_path=rel_csv_dir_path, rst_path=DOCS_DIR / "data_dictionaries/codes_and_labels.rst")
```
Very long line. Are the linters not catching anything > 88 characters?

katie-lamb replied:
There's something wrong with my pre-commit hooks maybe? Cornflakes was never catching lines that were too long but normally the pre-commit hook does. Will look into this.

zaneselvans replied:
I guess if we're going to have this linter run in the pre-commit hooks we should probably turn it on in our editors / Tox too. It'll just mean there's a big backlog of long lines that need to be formatted.

Maybe we should just switch over to using `black`. All this formatting stuff is so tedious.

zaneselvans commented on src/pudl/metadata/classes.py:
```diff
@@ -529,6 +536,22 @@ def from_id(cls, x: str) -> 'Encoder':
+        """Ouput dataframe to a csv for use in jinja template. Then output to an RST file."""
+        abs_csv_filepath = Path(abs_csv_dir_path, f"{self.name}.csv")
+        self.df.to_csv(abs_csv_filepath, index=False)
+        rel_csv_filepath = Path(rel_csv_dir_path, f"{self.name}.csv")
+        description = RESOURCE_METADATA[self.name]["description"]
```
Instead of creating these single-use variables which are only used a single time in `template.render()` below, I find it easier to read what's happening with the values themselves as arguments. It saves one step of mental redirection. E.g.

```py
self.df.to_csv(
    abs_csv_dir_path / f"{self.name}.csv",
    index=False,
)

return template.render(
    Encoder=self,
    description=RESOURCE_METADATA[self.name]["description"],
    csv_filepath=rel_csv_dir_path / f"{self.name}.csv",
    is_header=is_header
)
```

zaneselvans reviewed (changes requested):
I made several comments, but the only thing I think really needs to get changed is the formatting of the RST output -- I think we should make it match the existing PUDL DB data dictionary insofar as we can, with each coding table getting a sub-header for direct linking, sidebar navigation, and more separation between adjacent sections.

katie-lamb commented on src/pudl/metadata/classes.py:
```diff
@@ -529,6 +536,22 @@ def from_id(cls, x: str) -> 'Encoder':
+
+    def to_rst(self, abs_csv_dir_path: DirectoryPath, rel_csv_dir_path: DirectoryPath, is_header: Bool) -> String:
+        """Ouput dataframe to a csv for use in jinja template. Then output to an RST file."""
+        abs_csv_filepath = Path(abs_csv_dir_path, f"{self.name}.csv")
```
It's all a little tedious because the `csv-table` directive wants an absolute path that is relative to the source directory but pandas `to_csv` wants an absolute path. I tried to clean this up a little using `top_dir` and `csv_subdir`

zaneselvans commented on src/pudl/metadata/templates/codemetadata.rst.jinja:
```diff
@@ -0,0 +1,33 @@
+Also included is a table of non-standard codes that are mapped to canonical, standardized codes.
+{% endif %}
+
+.. _{{ Encoder.name }}:
```
I could be wrong but I think [explicit targets like this](https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html#explicit-targets) need to be globally unique within the documentation, and since we're using the table name already in the PUDL DB data dictionary, there's going to be a collision here. I'm surprised Sphinx didn't complain about this during the docs build.

Ah, looks like [it threw some warnings](https://readthedocs.org/api/v2/build/15740751.txt) but not errors in the RTD build. Did you try building the docs locally with `tox -e docs`?

Do you have an account on Read The Docs? If so I can add you as a maintainer so you can see the builds (if they aren't public). I tried to add `katie-lamb` but it said that account didn't exist.

```
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:207: WARNING: duplicate label contract_types_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:267: WARNING: duplicate label energy_sources_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:315: WARNING: duplicate label entity_types_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:527: WARNING: duplicate label fuel_transportation_modes_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:554: WARNING: duplicate label fuel_types_aer_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:2233: WARNING: duplicate label prime_movers_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
/home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/pudl_db.rst:2354: WARNING: duplicate label sector_consolidated_eia, other instance in /home/docs/checkouts/readthedocs.org/user_builds/catalystcoop-pudl/checkouts/1388/docs/data_dictionaries/codes_and_labels.rst
```

katie-lamb replied:
Yeah, I was experimenting with either suppressing warnings or adding a prefix but that wasn't working. I pushed it because my `tox -e docs` was having an issues before so I wasn't sure if it was just on my machine.

katie-lamb replied:
Ah will read more about explicit targets and suffixes. Probably better than suppressing the warning

zaneselvans reviewed (changes requested):
Docs build warnings due to duplicate cross reference targets. I think you can just add a suffix to the reference.

Any idea what the heck broke all the builds? It happened on the dependabot PR too. :-(
